frontend/src/main/java/com/google/tripmeout/frontend/storage/PersistentPlaceVisitStorage.java

```
@@ -0,0 +1,140 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyFactory.Builder;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.repackaged.com.google.datastore.v1.Datastore;
+import com.google.inject.Inject;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+import com.google.tripmeout.frontend.error.TripMeOutException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+class PersistentPlaceVisitStorage implements PlaceVisitStorage {
+  private static final String TRIP_ID_PROPERTY_NAME = "tripId";
+  private static final String USER_MARK_PROPERTY_NAME = "userMark";
+  private static final String ID_PROPERTY_NAME = "id";
+  private static final String PLACE_NAME_PROPERTY_NAME = "placeName";
+  private static final String PLACE_API_ID_PROPERTY_NAME = "placeApiId";
+
+  private final DatastoreService datastore;
+
+  @Inject
+  PersistentPlaceVisitStorage(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyExistsException {
+    Key placeKey = KeyFactory.stringToKey(placeVisit.id());
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      datastore.get(placeKey);
+      transaction.rollback();
+      throw new PlaceVisitAlreadyExistsException("Place Already Exists");
+    } catch (EntityNotFoundException e) {
```

Please add a comment to this empty catch block (and elsewhere).